copt/randomized.py

```diff
@@ -691,3 +693,149 @@ def epoch_iteration_template(
             memory_gradient[i] = grad_i
 
     return epoch_iteration_template
+
+
+def step_size_sfw(variant):
+    if variant == 'SAG':
+        def step_sizes_SAG(t, n_samples=None):
+            step_size_x = 2 / (t+2)
+            return step_size_x, None
+        return step_sizes_SAG
+
+    if variant == 'MK':
+        def step_sizes_MK(t, n_samples=None):
+            step_size_x = 1 / (t+1)
+            step_size_agg = step_size_x ** (2/3)
+            return step_size_x, step_size_agg
+        return step_sizes_MK
+
+    if variant == 'LF':
+        def step_sizes_LF(t, n_samples=None):
+            if n_samples is None:
+                raise ValueError("n_samples must be the number of samples in the dataset.")
+            step_agg = 2 * n_samples / (2 * n_samples + t + 1)
+            step_x = 2 * (2 * n_samples + t) / ((t+1) * (4 * n_samples + t + 1))
+            return step_x, step_agg
+        return step_sizes_LF
+
+
+def minimize_sfw(
+        f_deriv,
+        A,
+        b,
+        x0,
+        lmo,
+        step_size=None,
+        max_iter=500,
+        tol=1e-6,
+        verbose=False,
+        callback=None,
+        variant='SAG'
+):
+    r"""Stochastic Frank-Wolfe (SFW) algorithm.
+
+    The SFW algorithm can solve optimization problems of the form
+
+        argmin_{x \in constraint} (1/n)\sum_{i}^n_samples f(A_i^T x, b_i)
+
+    Args:
+      f_deriv
+          derivative of f
+
+      x0: np.ndarray
+          Starting point for optimization.
+
+      step_size: function or None, optional
+          Step size for the optimization. If None is given, this will be set as the
+          default: t -> 2/(t+2)
+
+      lmo: function
+          returns the update direction
+
+      max_iter: int
+          Maximum number of gradient calls in the optimization.
+
+      tol: float
+          Tolerance criterion. The algorithm will stop whenever the
+          difference between two successive iterates is below tol.
+
+      verbose: bool
+          Verbosity level. True might print some messages.
+
+      callback: function or None
+          If not None, callback will be called at each iteration.
+
+      variant: str in {'SAG', 'MK', 'LF'}
+          Controls which variant of SFW to use.
+          'SAG' is described in Negiar et al. (2020),
+          'MK' in Mokhtari et al. (2020),
+          'LF' in Lu and Freund (2020).
+
+    Returns:
+      opt: OptimizeResult
+          The optimization result represented as a
+          ``scipy.optimize.OptimizeResult`` object. Important attributes are:
+          ``x`` the solution array, ``success`` a Boolean flag indicating if
+          the optimizer exited successfully and ``message`` which describes
+          the cause of the termination. See `scipy.optimize.OptimizeResult`
+          for a description of other attributes.
+
+    """
```

Comment:
Please add the appropriate references, as in **minimize_frank_wolfe**

Reply:
(for the SAG version reference our arxiv v2)

Reply:
Done